uarray/_uarray_dispatch.cxx

```diff
@@ -0,0 +1,1044 @@
+class context_helper
+{
+  T new_backend_;
+  std::vector<T> * backends_;
+  size_t enter_size_;
+public:
+
+  context_helper():
+    backends_(nullptr),
+    enter_size_(size_t(-1))
+    {}
+
+  bool init(std::vector<T> & backends, T new_backend)
+    {
+      backends_ = &backends;
+      new_backend_ = std::move(new_backend);
+      return true;
+    }
+
+  bool enter()
+    {
+      enter_size_ = backends_->size();
+      try { backends_->push_back(new_backend_); }
+      catch(std::bad_alloc&)
+      {
+        PyErr_NoMemory();
+        return false;
+      }
+      return true;
+    }
+
+  bool exit()
+    {
+      bool success = (enter_size_ + 1 == backends_->size()
+                      && backends_->back() == new_backend_);
+      if (enter_size_ < backends_->size())
+        backends_->resize(enter_size_);
+
+      if (!success)
+        PyErr_SetString(PyExc_RuntimeError,
+                        "Found invalid context state while in __exit__");
+      return success;
+    }
+};
+
+
+struct SetBackendContext
+{
+  PyObject_HEAD
+
+  context_helper<backend_options> ctx_;
+
+  static void dealloc(SetBackendContext * self)
+    {
+      self->~SetBackendContext();
+      Py_TYPE(self)->tp_free(self);
+    }
+
+  static PyObject * new_(PyTypeObject * type, PyObject * args, PyObject * kwargs)
+    {
+      auto self = reinterpret_cast<SetBackendContext *>(type->tp_alloc(type, 0));
+      if (self == nullptr)
+        return nullptr;
+
+      // Placement new
+      self = new (self) SetBackendContext;
+      return reinterpret_cast<PyObject *>(self);
+    }
+
+  static int init(
+    SetBackendContext * self, PyObject * args, PyObject * kwargs)
+    {
+      static const char * kwlist[] = {"backend", "coerce", "only", nullptr};
+      PyObject * backend = nullptr;
+      PyObject * coerce = nullptr;
+      PyObject * only = nullptr;
+
+      if (!PyArg_ParseTupleAndKeywords(
+            args, kwargs,
+            "O|O!O!", (char**)kwlist,
+            &backend,
+            &PyBool_Type, &coerce,
+            &PyBool_Type, &only))
+        return -1;
+
+
+      auto domain = backend_to_domain_string(backend);
+      if (domain.empty())
+        return -1;
+      backend_options opt;
+      opt.backend = py_ref::ref(backend);
+      opt.coerce = (coerce == Py_True);
+      opt.only = (only == Py_True);
+
+      try
+      {
+        if (!self->ctx_.init(local_domain_map[domain].preferred, std::move(opt)))
+          return -1;
+      }
+      catch(std::bad_alloc&)
+      {
+        PyErr_NoMemory();
+        return -1;
+      }
+
+      return 0;
+    }
+
+  static PyObject * enter__(SetBackendContext * self, PyObject * /*args*/)
+    {
+      if (!self->ctx_.enter())
+        return nullptr;
+      Py_RETURN_NONE;
+    }
+
+  static PyObject * exit__(SetBackendContext * self, PyObject * /*args*/)
+    {
+      if (!self->ctx_.exit())
+        return nullptr;
+      Py_RETURN_NONE;
+    }
+};
+
+
+struct SkipBackendContext
```

Comment:
Would it make the code simpler to have an abstract class and reduce the duplication?

Reply:
`context_helper` already covers most of the duplication. `SetBackendContext` and `SkipBackendContext` have slightly different arguments and data types so do need to be distinct.

I could maybe add something like a `py_object` base class that covers `new`, `dealloc` and holds `PyObject_HEAD`. Does that sound useful?

Reply:
I tried creating a base class and while it did seem to work, I think it's probably a bad idea to try and mix C++ inheritance with the Python c-style "inheritance". It makes very low level assumptions about the memory layout of types that don't necessarly hold for C++ types. I also get this warning that because of it:
```
uarray/_uarray_dispatch.cxx:846:12: warning: offsetof within non-standard-layout type ‘{anonymous}::Function’ is undefined [-Winvalid-offsetof]
   offsetof(Function, dict_),      /* tp_dictoffset */
```